core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala

```diff
@@ -0,0 +1,252 @@
+  extends OutputCommitCoordinationMessage
+private[spark] case class StageEnded(stage: Int) extends OutputCommitCoordinationMessage
+private[spark] case object StopCoordinator extends OutputCommitCoordinationMessage
+
+private[spark] case class AskPermissionToCommitOutput(
+    stage: Int,
+    task: Long,
+    partId: Int,
+    taskAttempt: Long)
+  extends OutputCommitCoordinationMessage with Serializable
+
+private[spark] case class TaskCompleted(
+    stage: Int,
+    task: Long,
+    partId: Int,
+    attempt: Long,
+    successful: Boolean)
+  extends OutputCommitCoordinationMessage
+
+/**
+ * Authority that decides whether tasks can commit output to HDFS.
+ *
+ * This lives on the driver, but the actor allows the tasks that commit
+ * to Hadoop to invoke it.
+ */
+private[spark] class OutputCommitCoordinator(conf: SparkConf) extends Logging {
+
+  private type StageId = Int
+  private type PartitionId = Int
+  private type TaskId = Long
+  private type TaskAttemptId = Long
+
+  // Wrapper for an int option that allows it to be locked via a synchronized block
+  // while still setting option itself to Some(...) or None.
+  private class LockableAttemptId(var value: Option[TaskAttemptId])
+
+  private type CommittersByStageHashMap =
+    ConcurrentHashMap[StageId, ScalaImmutableMap[PartitionId, LockableAttemptId]]
+
+  // Initialized by SparkEnv
+  private var coordinatorActor: Option[ActorRef] = None
+  private val timeout = AkkaUtils.askTimeout(conf)
+  private val maxAttempts = AkkaUtils.numRetries(conf)
+  private val retryInterval = AkkaUtils.retryWaitMs(conf)
+  private val authorizedCommittersByStage = new CommittersByStageHashMap().asScala
+
+  private var executorRequestHandlingThreadPool: Option[ExecutorService] = None
+
+  def stageStart(stage: StageId, partitionIds: Seq[Int]): Unit = {
+    sendToActor(StageStarted(stage, partitionIds))
+  }
+
+  def stageEnd(stage: StageId): Unit = {
+    sendToActor(StageEnded(stage))
+  }
+
+  def canCommit(
+      stage: StageId,
+      task: TaskId,
+      partId: PartitionId,
+      attempt: TaskAttemptId): Boolean = {
+    askActor(AskPermissionToCommitOutput(stage, task, partId, attempt))
+  }
+
+  def taskCompleted(
+      stage: StageId,
+      task: TaskId,
+      partId: PartitionId,
+      attempt: TaskAttemptId,
+      successful: Boolean): Unit = {
+    sendToActor(TaskCompleted(stage, task, partId, attempt, successful))
+  }
+
+  def stop(): Unit = {
+    executorRequestHandlingThreadPool.foreach { pool =>
+      pool.shutdownNow()
+      pool.awaitTermination(10, TimeUnit.SECONDS)
+    }
+    sendToActor(StopCoordinator)
+    coordinatorActor = None
+    executorRequestHandlingThreadPool = None
+    authorizedCommittersByStage.clear
+  }
+
+  def initialize(actor: ActorRef, isDriver: Boolean): Unit = {
+    coordinatorActor = Some(actor)
+    executorRequestHandlingThreadPool = {
+      if (isDriver) {
+        Some(Executors.newFixedThreadPool(4))
```

Comment:
I would be surprised if submitting the job took longer than the job itself. The time spent to compute the canCommit computation should be at least the same amount of time to submit the job. Given this I think the parallelism will help. @vanzin what do you think?